packages/vite/src/node/server/persistentCache.ts

```diff
@@ -0,0 +1,443 @@
+}
+
+export interface PersistentCacheResult {
+  code: string
+  map?: any
+}
+
+export interface PersistentCacheFile {
+  relatedModules: Record<string, string>
+}
+
+export async function createPersistentCache(
+  config: ResolvedConfig
+): Promise<PersistentCache | null> {
+  const options = config.serverPersistentCache
+  if (!options) {
+    return null
+  }
+
+  const logger = config.logger
+  const debugLog = createDebugger('vite:persistent-cache')
+
+  // Cache directory
+
+  const resolvedCacheDir = normalizePath(path.resolve(options.cacheDir))
+
+  if (!fs.existsSync(resolvedCacheDir)) {
+    fs.mkdirSync(resolvedCacheDir, { recursive: true })
+  }
+
+  // Cache version
+
+  const hashedVersionFiles = await Promise.all(
+    options.cacheVersionFromFiles.map((file) => {
+      if (!fs.existsSync(file)) {
+        throw new Error(`Persistent cache version file not found: ${file}`)
+      }
+      return fs.promises.readFile(file, 'utf-8')
+    })
+  ).then((codes) => getCodeHash(codes.join('')))
+
+  const defineHash = config.define
+    ? getCodeHash(JSON.stringify(config.define))
+    : ''
+
+  const envHash = getCodeHash(JSON.stringify(config.env))
+
+  const cacheVersion = [
+    options.cacheVersion,
+    `vite:${version}`,
+    config.configFileHash,
+    hashedVersionFiles,
+    defineHash,
+    envHash
+  ]
+    .filter(Boolean)
+    .join('-')
+
+  // Manifest
+
+  const manifestPath = path.join(resolvedCacheDir, 'manifest.json')
+  let manifest: PersistentCacheManifest | null = null
+  if (fs.existsSync(manifestPath)) {
+    try {
+      manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf-8'))
+      if (manifest && manifest.version !== cacheVersion) {
+        // Bust cache if version changed
+        debugLog(
+          `Clearing persistent cache (${cacheVersion} from ${manifest.version})...`
+        )
+        try {
+          // Empty the directory
+          const files = await fs.promises.readdir(resolvedCacheDir)
+          await Promise.all(
+            files.map((file) =>
+              fs.promises.unlink(path.join(resolvedCacheDir, file))
+            )
+          )
+          logger.info(`Deleted ${files.length} files.`)
+        } catch (e) {
+          logger.warn(
+            colors.yellow(
+              `Failed to empty persistent cache directory '${resolvedCacheDir}': ${e.message}`
+            )
+          )
+        }
+        manifest = null
+      }
+    } catch (e) {
+      logger.warn(
+        colors.yellow(
+          `Failed to load persistent cache manifest '${manifestPath}': ${e.message}`
+        )
+      )
+    }
+  }
+  const resolvedManifest: PersistentCacheManifest = manifest ?? {
+    version: cacheVersion,
+    modules: {},
+    files: {}
+  }
+
+  // Manifest write queue
+
+  let isManifestWriteQueued = false
+  let isManifestWriting = false
+  let manifestWriteTimer: any = null
+
+  function queueManifestWrite() {
```

Comment:
Does this do same thing? Easier to grok?

```typescript
const queueManifestWrite = (function() {
  let timer: NodeJS.Timeout | undefined
  let writing = false
  let rewrite = false
  const write = async () => {
    timer = undefined
    writing = true
    // do write with async stuff and try/catch
    writing = false
    if (rewrite) {
      rewrite = false
      write()
    }
  }
  return () => {
    if (writing) {
      rewrite = true
      return
    }
    if (timer) {
      return
    }
    timer = setTimeout(write, 1000)
  }
})()
```

Reply:
I'm not sure :cat2: